lib/pure/asyncdispatch.nim

```diff
@@ -262,24 +277,121 @@ when defined(windows) or defined(nimdoc):
     PostCallbackDataPtr = ptr PostCallbackData
 
     AsyncEventImpl = object
-      hEvent: Handle
-      hWaiter: Handle
-      pcd: PostCallbackDataPtr
+      triggered: bool
+      when compileOption("threads"):
+        eventLock: RLock
+      p: PDispatcher
+      vFD: VirtualFD
+      callbacks: seq[Callback]
```

Comment:
This is not threadsafe. ``seq`` are thread-local.

Reply:
The callback seq should only ever be accessed by a single thread: the thread that owns the async event loop.

It would be very odd for a thread to register a callback on a different threads event loop. Even if this were made this thread safe, the callback itself might access memory from a different thread.

Reply:
A couple of options:
- I could factor the seq out and create a global seq in the asyncDispatcher singleton.
- Or I could make it a single callback. I made this a seq to make it more extensible, but it's not required. (In fact, the old system only allows a single callback to be tied to an AsyncEvent).

Reply:
old AsyncEvent is able to be used between threads, but proposed AsyncEvent is unable to be used between threads.

Reply:
Cheatface, can you give an example? 

Old async events is not threadsafe. The old AsyncEvent allowed cross thread access to the asyncEvent memory with no synchronization or race protection at all. This system explicitly synchronizes threads through locking.

Reply:
> Old async events is not threadsafe. The old AsyncEvent allowed cross thread access to the asyncEvent memory with no synchronization or race protection at all. This system explicitly synchronizes threads through locking.

**Windows**:
Sorry but not only locks are used to synchronize threads: https://github.com/nim-lang/Nim/blob/devel/lib/pure/asyncdispatch.nim#L810 . Windows OS has events (which in Posix often called conditional variables).
**Linux**:
Linux version uses `eventfd` https://github.com/nim-lang/Nim/blob/devel/lib/pure/ioselects/ioselectors_epoll.nim#L112-L118.
So lets check about thread safety of **eventfd()** via `man eventfd`:
```
ATTRIBUTES
       For an explanation of the terms used in this section, see attributes(7).

       ┌──────────┬───────────────┬─────────┐
       │Interface │ Attribute     │ Value   │
       ├──────────┼───────────────┼─────────┤
       │eventfd() │ Thread safety │ MT-Safe │
       └──────────┴───────────────┴─────────┘
```
BSD/MacOS and all other Posix systems:
All this systems are using OS pipes:
https://github.com/nim-lang/Nim/blob/devel/lib/pure/ioselects/ioselectors_kqueue.nim#L136
https://github.com/nim-lang/Nim/blob/devel/lib/pure/ioselects/ioselectors_poll.nim#L191

And some information about pipes form `man pipe`:
```
pipe()  creates a pipe, a unidirectional data channel that can be used for interprocess communication.
```
So i think if its created for interprocess communication it will be fine to use it for interthread communication.

Reply:
This is example of use:
https://gist.github.com/cheatfate/87b672034be5b71ab6a975266558ed8c
But if you encounter any problems its some internal asyncdispatch bugs, i'm not involved in asyncdispatch development anymore so i'm not following it. But this sample is exactly shows how it must work.

Reply:
@cheatfate, your test actually fails on current devel Nim as follows:
```
/home/ray/Nim/Nim-devel/cheatfateAsync.nim(38) threadProc2
/home/ray/.choosenim/toolchains/nim-#devel/lib/pure/asyncdispatch.nim(1879) waitFor
/home/ray/.choosenim/toolchains/nim-#devel/lib/pure/asyncdispatch.nim(1569) poll
/home/ray/.choosenim/toolchains/nim-#devel/lib/pure/asyncdispatch.nim(1335) runOnce
/home/ray/.choosenim/toolchains/nim-#devel/lib/pure/asyncdispatch.nim(210) processPendingCallbacks
/home/ray/.choosenim/toolchains/nim-#devel/lib/pure/asyncmacro.nim(34) asyncProc2NimAsyncContinue
/home/ray/Nim/Nim-devel/cheatfateAsync.nim(28) asyncProc2Iter
/home/ray/Nim/Nim-devel/cheatfateAsync.nim(17) wait
/home/ray/.choosenim/toolchains/nim-#devel/lib/pure/asyncdispatch.nim(1556) addEvent
/home/ray/.choosenim/toolchains/nim-#devel/lib/pure/ioselects/ioselectors_epoll.nim(368) registerEvent
/home/ray/.choosenim/toolchains/nim-#devel/lib/pure/selectors.nim(279) raiseIOSelectorsError
[[reraised from:
/home/ray/Nim/Nim-devel/cheatfateAsync.nim(38) threadProc2
/home/ray/.choosenim/toolchains/nim-#devel/lib/pure/asyncdispatch.nim(1881) waitFor
/home/ray/.choosenim/toolchains/nim-#devel/lib/pure/asyncfutures.nim(383) read
]]
Error: unhandled exception: File exists (code: 17)
Async traceback:
  /home/ray/Nim/Nim-devel/cheatfateAsync.nim(38)                                           threadProc2
  /home/ray/.choosenim/toolchains/nim-#devel/lib/pure/asyncdispatch.nim(1879)              waitFor
  /home/ray/.choosenim/toolchains/nim-#devel/lib/pure/asyncdispatch.nim(1569)              poll
  /home/ray/.choosenim/toolchains/nim-#devel/lib/pure/asyncdispatch.nim(1335)              runOnce
  /home/ray/.choosenim/toolchains/nim-#devel/lib/pure/asyncdispatch.nim(210)               processPendingCallbacks
  /home/ray/.choosenim/toolchains/nim-#devel/lib/pure/asyncmacro.nim(34)                   asyncProc2NimAsyncContinue
  /home/ray/Nim/Nim-devel/cheatfateAsync.nim(28)                                           asyncProc2Iter
  /home/ray/Nim/Nim-devel/cheatfateAsync.nim(17)                                           wait
  /home/ray/.choosenim/toolchains/nim-#devel/lib/pure/asyncdispatch.nim(1556)              addEvent
  /home/ray/.choosenim/toolchains/nim-#devel/lib/pure/ioselects/ioselectors_epoll.nim(368) registerEvent
  /home/ray/.choosenim/toolchains/nim-#devel/lib/pure/selectors.nim(279)                   raiseIOSelectorsError
Exception message: File exists (code: 17)
Exception type: [IOSelectorsException]
Error: execution of an external program failed: '/home/ray/Nim/Nim-devel/cheatfateAsync '
```

On the VirtualAsyncPR it does not have this issue but hits a race condition between the `addEvent` and `trigger` procs.
line 17 and lines 22/28 respectively.

It's possible to trigger the event in one thread, before the other thread has "re-registered" the event. I'm not sure if AsyncEvent was meant to be constantly unregistered and re-registered this way.

Reply:
The following change fixes the race condition I observe with my PR: https://gist.github.com/rayman22201/e6a9dcde22e009dbd397b8755eae93d7#file-asyncevent-nim-L22
That guarantees the event is registered before it is triggered on either thread.

Note: the test still fails with the `File exists (code: 17)` Exception on Devel.